### ISSUE TITLE
Monorepo structure

### ISSUE THREAD

**Issue by tomlinton** [OriginProtocol/origin-devops#30]
Following on from #16 and the latest engineering call I wanted to kick off some discussion about the structure of the potential monorepo. @nick had some ideas about the [directory structure ](https://github.com/OriginProtocol/origin-devops/issues/16#issuecomment-409942378) which were great and I wanted to propose a similar structure that I think would work well with `origin-box` and `docker-compose`:

```
- dockerfiles
- envfiles
- origin-dapp
- origin-js
- origin-bridge
- origin-docs
- tests
- docker-compose.yml
- docker-compose-test.yml
- README
- ...
```
This is basically the directory structure used by `origin-box` at the moment. It makes sense for the Docker stuff to be at the top level if it is managing things. Some files ancillary to `origin-box` like the readme will need to find a new home as the top level readme should be for Origin as a whole.

The `tests` directory will eventually contain tests for the whole stack, for example we can use a container to mock the Twilio Verify API and test the whole attestation flow using a headless browser.

Some thoughts/discussion points:

- Unsure about the `origin-` prefix on all of the directory names. The main reason I've included it is because having a directory just called `js` is confusing. Perhaps `dapp`, `bridge`, and `origin-js`?
- Other necessary directories? @nick mentioned `playground`, `mobile` and `schemas`. @franckc might need a new one for the Apollo server?
- Haven't figured out the `origin-website` stuff. It should be a separate repo but I'm not sure if it's container still belongs here (it is part of `origin-box`).

Some future changes:
- I think it makes sense for the contracts to be moved to a top level directory (bought up by @nick in #16). Another container could be used to manage Ganache.
- Similarly the IPFS stuff should eventually live in a top level directory, `origin-bridge` and `origin-js` are now both doing IPFS things and it'd be great if a separate container could handle the projects IPFS requirements.

I mentioned in Discord that I'm a bit concerned about how invested we get in Docker. You can see above I'm proposing adding even more containers. :scream: I think the key for me is that if we go down the `docker-compose` road we might as well go all the way and have it managing everything, rather than having things spinning up their own bits and pieces (e.g IPFS servers). We definitely need some sort of solution to manage this complexity.

**Comment by micahalcorn** [comment]
Thanks for getting this started, @tomlinton. For what it's worth, we only recently renamed all of the repos to have an "origin-" prefix. That will be less important once they are nested in a single repo. Back to #16, I had proposed using tags as the method by which we promise stability to newcomers. We would also use that tag in the box so that people wouldn't even need to know what commits to use unless they want to venture out. I guess we can still do that with the containers in the same repo?

**Comment by DanielVF** [comment]
I think it's fine to keep the "origin-" prefixes. It makes it obvious which are the packages/apps.

**Comment by tomlinton** [comment]
Tags should work fine, we just tell people to checkout a specific tag when getting started. We probably would have to be careful the active development branch doesn't get too far ahead.

The `tests` directory in the root was intended for integration tests involving multiple components, e.g. testing `origin-dapp` and `origin-bridge` together rather than each in isolation (we don't have any of these yet). Each component would still have its own tests in its directory as it does now. It's not really important yet but down the track it might make an appearance.

**Comment by tomlinton** [comment]
I edited the directory structure in the first post to include some new things at the top level:

- indexing (the GraphQL server, currently in a branch of `origin-js`)
- event-listener (currently in a branch of `origin-js`)
- ipfs (anything IPFS related e.g. pinner, currently lives in `origin-devops`)
- orbit-db (currently lives in `origin-bridge`)

I think in general all the things that live at the top level should be services in the stack that need to be deployed. There are a few extra things that aren't services but still belong at the top level, I don't think we have to be strict about it, as long as services aren't embedded deep in subdirectories.

**Comment by DanielVF** [comment]
Maybe we could call the "indexing" service "origin-discovery"?

**Comment by DanielVF** [comment]
@micahalcorn If you need to tag releases of other services, you could just include the name of service in the tag name along with the version number.

**Comment by tomlinton** [comment]
@DanielVF I like the name! Perhaps something we should adopt even if we don't make it to a monorepo. I've just been calling it the GraphQL server which isn't very meaningful.

**Comment by franckc** [comment]
I like origin-discovery ! My only slight hesitation is that in the future the GraphQL server may support other features than just browse/search. Like for example mobile notifications. But I can't really come up with a better name. Maybe origin-services ? ... not great. naming is so hard ! :)

Related to naming, origin-bridge could probably get renamed to origin-attestation since that's the only thing the current bridge server will do for V1.

**Comment by nick** [comment]
I like origin-discovery too for indexing. I think if we have other GraphQL services such as messaging, we might consider having that in a separate dir (origin-messaging?) so we can keep the components opt-in. For example a partner might want messaging but not discovery, or vice versa.

**Comment by sparrowDom** [comment]
I would keep all the configuration in one `.env` file. Since we are transitioning to mono-repo I don't see why not have one "master" configuration file. IMHO there are some benefits to this approach: 
- different containers can share a variable. (for example ethNetwork: localhost:8545)
- should be less error prone, since you can inspect/set the whole deployment configuration from one file

**Comment by tomlinton** [comment]
@sparrowDom agreed it would be nice for convenience. We will lose a few things though:

- Ability to configure independently, e.g. to swap out an IPFS gateway for the DApp only to diagnose an issue
- Separate files provide some type of self-documentation since it is clear to developers what is configurable for each service

**Comment by DanielVF** [comment]
I'm now concerned that we are close enough to the public release that we shouldn't switch before it. Probably need to keep our focus on finishing and polishing.

**Comment by tomlinton** [comment]
I think you are right @DanielVF. There are a few simple things that we can do before the public release that would ease the transition and also help me out with deployments.

Right now `origin-box` is the closest thing to the directory structure we want for the monorepo so if we use that as a base and:

- Move the GraphQL/event listener out of `origin-js` and into the top level of `origin-box` and do the rename to `origin-discovery` or `origin-services`.
- Move the orbit-db messaging server out of `origin-bridge` and into the top level of `origin-box`. I can do this one because I have to do a bit of a restructure for deployment anyway.

If we have any additional services that are going to need deployment for the public release that don't really have a home (I know @franckc is working on a faucet, not sure where it lives at the moment) then it'd be great if they could be added to the top level of `origin-box`. I realise `origin-box` will then have a bunch of code in it rather than being just a development setup which is not ideal, but I think we can deal with it for a while.

**Comment by franckc** [comment]
@tomlinton See origin-js [PR #381](https://github.com/OriginProtocol/origin-js/pull/381) for the faucet code. It is currently lives under origin-js:/token. The only dependency it has on origin-js is the contracts JSON files. I can easily move it to a directory at the root of origin box (maybe origin-token ?). Same thing for the origin-js:/indexing, I can easily move it to origin-box:/origin-discovery.

**Comment by tomlinton** [comment]
That'd be great @franckc. Maybe call the faucet `origin-faucet` just to avoid any confusion with the token contracts?

**Comment by franckc** [comment]
Doc with mono-repo migration proposal: https://docs.google.com/document/d/1QBER5SbEqPtcMqk-J9fy7jV4xDqHoIirX0HSgqaJWMA/edit#
